core/metadata.jvm/src/org/jetbrains/kotlin/metadata/jvm/deserialization/ClassMapperLite.kt

```diff
@@ -13,6 +13,8 @@ package org.jetbrains.kotlin.metadata.jvm.deserialization
 // WARNING: improving the behavior of this class MAY BREAK BINARY COMPATIBILITY of code compiled by Kotlin, because it may make
 // the new compiler skip writing the signatures it now thinks are trivial, and the old compiler would recreate them incorrectly.
 object ClassMapperLite {
+    // Simply "kotlin", but to avoid being renamed by namespace relocation (e.g., Shadow.relocate gradle plugin)
+    private val kotlin = listOf('k', 'o', 't', 'l', 'i', 'n').joinToString(separator = "")
```

Comment:
Have you tried just `private val kotlin = "kotlin"` here?

Note that we had a change in compilation strategy and since 1.4, non-const vals should _not_ be inlined at call sites, and thus I don't think Shadow will rename the class names below. Unless of course it just renames any string literal "kotlin" found in the class file, which would be mind-boggling to say the least.

Reply:
I can't link an internal bug, but hopefully this code review has some clues: https://r8-review.googlesource.com/c/r8/+/23820

Anyhow, I'm pretty much sure Shadow renames any string literal "kotlin" if the relocation task is asked to do so: https://r8.googlesource.com/r8/+/refs/heads/master/build.gradle#701 and indeed, that was a surprise to us too. Actually, not just "kotlin", but literally any appearances, like "kotlin/Unit" -> "relocated_kotlin/Unit".

Reply:
Thanks, I've tried a small project with Shadow and it seems to be the case. Which looks like a bug in Shadow. At least, there could be a flag to disable relocation for string literals which are not used for reflection immediately.

Reply:
@madsager and I discussed this a couple times. Since other shrinkers also suffer from the same issue, it'd be good to avoid such plain prefix in general. However, it's also silly that that relocating tool is modifying literally all appearances of target string literals in the string pool. Maybe, `R8` (or actually `L8`, a library desugaring variant) has enough utils to do a better job, like reading `.jar`, building hierarchy, relocating target package (as it does for, e.g., `java.lang`), and outputting relocated `.jar`. I'll file a feature request after discussing other folks in `R8`. Meanwhile, hope this could be a reasonable workaround.